### ISSUE TITLE
History completion ideally should not be blocking

### ISSUE THREAD

**Issue by andrewcarlotti** [qutebrowser/qutebrowser#3989]
At the moment when filtering history for completion, any typos in the search term result in the full history file being loaded and searched before qutebrowser will respond further to input. For example, I frequently search for "simutrans", which would return results instantaneously because I have lots of results in very recent history. However, I frequently mistype this as "simtr...", at which point I have to wait for the full history to be searched before I can correct my mistake.

The last three times I tried this I had to wait 12 seconds (on version 1.3.1), and 40 seconds and then 10 seconds (on the version detailed below). My history file is 115MB; I don't know how much history this corresponds to. I am using a 6.5-year-old laptop which is becoming underpowered for various reasons. I think the variation in running times probably depends on how much of the file is in cache and how much of it needs to be read from the disk.

Better behaviour would be to have the search run in the background with the possibility of cancelling the search if the search term is deleted mid-search, and ideally with the ability to return and display partial results whilst still searching. I realise that some of this could be difficult/impossible with the current completion system, but surely some improvements can be made?

```
qutebrowser v1.3.2
Git commit: 
Backend: QtWebEngine (Chromium 65.0.3325.151)

CPython: 3.6.5
Qt: 5.11.0
PyQt: 5.10.1

sip: 4.19.8
colorama: 0.3.9
pypeg2: 2.15
jinja2: 2.10
pygments: 2.2.0
yaml: 3.12
cssutils: no
attr: 18.1.0
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebKitWidgets: yes
pdf.js: no
sqlite: 3.24.0
QtNetwork SSL: OpenSSL 1.1.0h  27 Mar 2018

Style: QFusionStyle
Platform: Linux-4.16.13-2-ARCH-x86_64-with-arch, 64bit
Linux distribution: Arch Linux (arch)
Frozen: False
Imported from /usr/lib/python3.6/site-packages/qutebrowser
Qt library executable path: /usr/lib/qt/libexec, data path: /usr/share/qt

Paths:
cache: /home/andrew/.cache/qutebrowser
config: /home/andrew/.config/qutebrowser
data: /home/andrew/.local/share/qutebrowser
runtime: /run/user/1000/qutebrowser
system data: /usr/share/qutebrowser

Uptime: 0:44:08
```

**Comment by jgkamat** [comment]
Hi Andrew

Andrew Carlotti writes:

> The last three times I tried this I had to wait 12 seconds (on version 1.3.1), and 40 seconds and then 10 seconds (on the version detailed below). My history file is 115MB; I don't know how much history this corresponds to. I am using a 6.5-year-old laptop which is becoming underpowered for various reasons. I think the variation in running times probably depends on how much of the file is in cache and how much of it needs to be read from the disk.
>

You may want to see #1099 and #3507 for issues related to the completion
performance. I use qb on several slow machines, and I don't see much slowdown
from the completion. Perhaps you have some very large URLs in your completion?
my history file is ~70mb.

It may be possible to make completion asynchronous, but that could introduce a
lot of bugs (and may not be worth it).

**Comment by user202729** [comment]
I don't think this issue should be closed. Even if the query performance can't be improved, it can still be made to run asynchronously (as jgkamat suggested above)

Currently, running the query block the UI because they're run on the same thread.

**Comment by The-Compiler** [comment]
It's probably hard to impossible because Qt GUI access needs to be done in the main thread, and Qt's SQL stuff AFAIK isn't designed to run separately from the GUI part. Reimplementing the SQL stuff in Python likely also will make it slower (and will still block the GUI, due to Python's GIL).

So if you claim this is somehow possible, I'm going to need more proof of that :wink:

**Comment by jgkamat** [comment]
Why exactly does the GIL remain locked while the sql query is running? The sql
client must be implemented in qt, where the GIL isn't an issue.

Maybe there's something we can do to make PyQt5 release the GIL for us while
this query is running.

**Comment by user202729** [comment]
I don't think the GIL is locked. Rather, the `setQuery` is executed synchronously in the main thread, so it hangs the main loop.

**Comment by The-Compiler** [comment]
PyQt seems to release the GIL while the query is running (`QSqlQuery.exec()` has a `/ReleaseGIL/` annotation at least). But AFAIK we aren't even sure whether it's sqlite itself hanging, or Qt when processing/rendering/... the results.

Either way, anything we do in Python other than the actual execution of the query probably will hold the GIL, and will be slower than letting Qt do it. qutebrowser's data model behind the completion used to be a custom QStandardItemModel (and before that, a custom QAbstractItemModel), so we've been through various iterations of improving things already. Implementing a custom model is quite error-prone, cumbersome, and likely to be rather slow (because there's are lot of performance-critical work here). I'm really not convinced going back to that (*especially* with multithreading thrown in) will make things any better.

So yeah, been there, done that. Only a working implementation which is actually faster (or feels faster due to being async) than what we have currently will convince me otherwise. However, that's going to be a lot of effort, and if it's crazy complex, or significantly slower than the synchronous access (I'd bet it will be both) I still reserve the right to say no :wink:

cc @rcorre

**Comment by user202729** [comment]
I implemented a simple program that call `set_pattern` in a separate thread. This program use `time.sleep` to simulate slow query. The database connection must be made in that thread.

http://0x0.st/zTDq.py

Possible improvement:

* It'd be great if it's somehow possible to cancel the currently-running call when a new one comes.
* Currently `data` calls are not cached. If that's a bottleneck then it should be improved as well.

**Comment by The-Compiler** [comment]
Oh, interesting, thanks @user202729! That's definitely worth some more investigation then...

**Comment by user202729** [comment]
Unfortunately, it's hard to terminate a running SQL query.

https://www.qtcentre.org/threads/23905-Abort-QSqlQuery-exec()
https://stackoverflow.com/questions/17636592/abort-interrupt-kill-long-running-cross-sql-select-query-in-qt5

**Comment by lifranc** [comment]
Now that my history DB is quite large, typing urls is a f**in pain, basically unusable. My first instinct is to delete it, but I'd rather trim old entries. Does anyone have a script to do so? Might be a useful command to integrate in qutbrowser...

**Comment by ninewise** [comment]
`:set completion.web_history.max_items` helps to search only more recent entries.

The delay seems like a nice idea, though.

**Comment by The-Compiler** [comment]
You mean, like, `completion.delay`? :wink:

See #58 for deleting things on shutdown, and there's also `completion.web_history.exclude` which takes URL patterns which won't be added to the completion.
